rfd/0210-windows-ad-hsm.md

```diff
@@ -0,0 +1,221 @@
+and MS SQL database access. In both cases, Teleport generates an LDAP URL based
+on the configured Active Directory domain. It takes the form:
+
+```
+ldap://CN=CLUSTER,CN=SIGNER,CN=CDP,CN=Public Key Services,CN=Configuration,DC=example,DC=com
+```
+
+Where:
+- `CLUSTER` is the name of the Teleport cluster
+- `SIGNER` is `Teleport` (for Windows desktop certs) or `TeleportDB` for database certs
+- `example.com` is the AD domain (`DC=example,DC=com` in LDAP-speak)
+
+### CRL Publishing
+
+CRL publishing works differently for each Teleport feature.
+
+#### MS SQL database access
+
+For database access, our setup instructions require that users generate the CRL
+with `tctl auth crl`, copy it over to the Windows environment, and publish it to
+LDAP using `certutil -dspublish`.
+
+It is worth noting that `tctl auth crl` generates empty CRLs that are valid for _1 year_,
+so database access will require manual intervention in case:
+
+- the CRL expires after 1 year
+- a CA rotation is performed
+
+#### Windows desktop access
+
+For Windows desktop access, Teleport's `windows_desktop_service` generates a CRL
+on startup and publishes it to LDAP directly - no `certutil -dspublish` step is
+necessary. This process is repeated every time the agent restarts, as well as
+every time the agent refreshes its credentials (typically every 4 hours or so).
+
+As a result, desktop access requires manual intervention in the case of a CA rotation,
+but does not suffer from the CRL expiry issue, since the CRL is constantly getting
+pushed back by 1 year.
+
+### Proposed Changes
+
+In order to support HSMs, which result in multiple active signers being present in the cluster,
+we will need to:
+
+1. Generate certs with a CDP that references the specific cert issuer
+2. Publish multiple CRLs to LDAP instead of a single CRL
+
+#### Cert Generation: use SKID of the issuer
+
+When generating certificates, Teleport will use the subject key ID of the signer to
+generate a CDP that is unique to that signer.
+
+```
+ldap://CN=IV4GC3LQNRSSAVDFNRSXA33SORJUWSKE,CN=Teleport,CN=CDP,CN=Public Key Services,CN=Services,CN=Configuration,DC=example,DC=com
+```
+
+In this case, `IV4GC3LQNRSSAVDFNRSXA33SORJUWSKE` is Base32-encoded SKID of the issuer
+
+#### Publishing: `tctl auth crl --out`
+
+For `tctl` we will follow the precedent set in
+[#51298](https://github.com/gravitational/teleport/pull/51298) for `tctl auth
+export`, and add a `--out` flag to `tctl auth crl`. This flag will cause `tctl`
+to write the CRL(s) (whether there is one or many) to disk rather than to emit a
+single CRL to standard out.
+
+Teleport users setting up MS SQL access will then be able to perform a setup
+procedure very similar to what they already do, they will only have to repeat
+the `certutil -dspublish` command  once for each individual CRL.
+
+TODO: in order to publish to auth server, the CN of the CRL would need to change.
+
+#### Publishing: add CRLs to the backend `certificate_authority`
```

Comment:
I would just like to see some description of when the auth server generates the CRLs and adds them to the certificate_authority in this design

maybe on auth startup the auth checks each key in each CA that needs CRLs. if it can sign with that key, and there is currently no CRL or the CRL expires in <1 year, it generates a new one and updates the CA in the backend. this will handle adding CRLs to existing CAs without them, and refreshing them semi-regularly

during CA rotation when generating new keys it should also generate and store the CRL

not sure if we'd want to also periodically check/refresh the CRLs or if just at auth startup is enough